### PR TITLE
NS linear solver Physics

### PR DESCRIPTION
**Pull request by GiudGiud**
wanted to see how hard it would be. Seems it wont be bad at all

left todo:
- [x] match all the tests for segregated-linear
- [x] add momentum gravity, only term with existing kernels

### CODE REVIEW

File: modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/linear-segregated/lid-driven-segregated-physics.i [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    ymin = 0
+    ymax = .1
+    nx = 3
+    ny = 3
+  []
+[]
+
+[Problem]
+  linear_sys_names = 'u_system v_system pressure_system'
+  previous_nl_solution_required = true
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/linear-segregated/lid-driven-segregated-physics.i:20]:
```suggestion
```

File: modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+    for (const auto d : make_range(dimension()))
+    {
+      if (gravity_vector(d) != 0)
+      {
+        params.set<MooseFunctorName>("source_density") = std::to_string(gravity_vector(d));
+        params.set<NonlinearVariableName>("variable") = _velocity_names[d];
+
+        getProblem().addFVKernel(kernel_type, kernel_name + NS::directions[d], params);
+      }
+    }
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C:252]:
```suggestion
      if (gravity_vector(d) != 0)
      {
        params.set<MooseFunctorName>("source_density") = std::to_string(gravity_vector(d));
        params.set<NonlinearVariableName>("variable") = _velocity_names[d];

        getProblem().addFVKernel(kernel_type, kernel_name + NS::directions[d], params);
      }
```

File: modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    else if (momentum_wall_type == "slip")
+    {
+      // Do nothing
+    }
+    else if (momentum_wall_type == "symmetry")
+    {
+      // Placeholder
+      mooseError("Unsupported boundary condition type: " + std::string(momentum_wall_type));
+    }
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C:420]:
```suggestion
    else 
      mooseError("Unsupported wall boundary condition type: " + std::string(momentum_wall_type));
```

Comment (grmnptr) [reply modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C:420]: Could be a param error too.

File: modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+[Problem]
+  #TODO: remove both
+  linear_sys_names = 'u_system v_system pressure_system'
+  previous_nl_solution_required = true
+[]
+
+[Physics]
+  [NavierStokes]
+    [FlowLinear/flow]
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:24]: Could be FlowSegregated(Linear). We will remove the old segregated capabilities once everything has been migrated to the new one. Or something similar.

This is just a conversation starter, the name could probably stay.

File: modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    dim = 2
+    dx = '0.3'
+    dy = '0.3'
+    ix = '3'
+    iy = '3'
+  []
+[]
+
+[Problem]
+  #TODO: remove both
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:17]: Yeah would be cool if we could do this under the hood. I think the previous solution flag is already in your code.

File: modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/linear-segregated/lid-driven-segregated-physics.i [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    ny = 3
+  []
+[]
+
+[Problem]
+  linear_sys_names = 'u_system v_system pressure_system'
+[]
+
+
+[Physics]
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/linear-segregated/lid-driven-segregated-physics.i:23]: Nice syntax!

**Review by grmnptr — CHANGES_REQUESTED** [review]
Pretty neat!

File: modules/navier_stokes/src/physics/WCNSFVFlowPhysicsBase.C [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+  // Initialization parameters
+  params.transferParam<std::vector<FunctionName>>(NSFVBase::validParams(), "initial_velocity");
+  params.transferParam<FunctionName>(NSFVBase::validParams(), "initial_pressure");
+
+  // Spatial discretization scheme
+  // Specify the numerical schemes for interpolations of velocity and pressure
+  params.transferParam<MooseEnum>(NSFVBase::validParams(), "velocity_interpolation");
+  params.transferParam<MooseEnum>(NSFVBase::validParams(), "momentum_advection_interpolation");
+  params.transferParam<bool>(NSFVBase::validParams(), "pressure_two_term_bc_expansion");
+  params.transferParam<Real>(INSFVMomentumAdvection::validParams(), "characteristic_speed");
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVFlowPhysicsBase.C:74]: ok I ll suppress

[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVFlowPhysicsBase.C:74]: or move the transfer

File: modules/navier_stokes/src/physics/WCNSFVFlowPhysicsBase.C [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+    params.set<FunctionName>("function") = getParam<FunctionName>("initial_pressure");
+
+    getProblem().addInitialCondition("FunctionIC", prefix() + _pressure_name + "_ic", params);
+  }
+}
+
+unsigned short
+WCNSFVFlowPhysicsBase::getNumberAlgebraicGhostingLayersNeeded() const
+{
+  unsigned short ghost_layers = 2;
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVFlowPhysicsBase.C:350]: that's great. I ll override

File: modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C (hunk 3 above)
Comment (GiudGiud) [reply modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C:420]: slip does not work?

File: modules/navier_stokes/doc/content/source/physics/WCNSFVLinearFlowPhysics.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+The kernels created for free flow for the pressure correction equation:
+
+- [LinearFVAnisotropicDiffusion.md] for the pressure diffusion term
+- [LinearFVDivergencen.md] for the pressure divergence term
+
+
+## Automatically defined variables
+
+The `WCNSFVLinearFlowPhysics` automatically sets up the variables which are
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/doc/content/source/physics/WCNSFVLinearFlowPhysics.md:40]: good point. will do

File: modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  linear_sys_names = 'u_system v_system pressure_system'
+  previous_nl_solution_required = true
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:19]:
```suggestion
```

File: modules/navier_stokes/include/physics/WCNSFVLinearFlowPhysics.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  virtual void addINSOutletBC() override;
+  virtual void addINSWallsBC() override;
+
+  virtual bool hasForchheimerFriction() const override { return false; };
+
+  virtual void addRhieChowUserObjects() override;
+
+  UserObjectName rhieChowUOName() const override;
+
+  /// Name of the vector to hold pressure momentum equation contributions
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/include/physics/WCNSFVLinearFlowPhysics.h:62]: removed

File: modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i (hunk 5 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:17]: on it

File: modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i (hunk 4 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:23]: changed. FlowLinear was weird.

File: modules/navier_stokes/doc/content/source/physics/WCNSFVLinearFlowPhysics.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/navier_stokes/doc/content/source/physics/WCNSFVLinearFlowPhysics.md:40]: done

File: modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i (hunk 5 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:17]: actually this isnt easy. I dont think the problem should be created by the Physics, and this line creates the sytems. we would need to add APIs to make the problem create new systems from the Physics

File: framework/src/physics/PhysicsBase.C [hunk 12]
```diff
@@ -276,6 +276,21 @@ PhysicsBase::copyVariablesFromMesh(const std::vector<VariableName> & variables_t
...
+  if (_problem->hasVariable(var_name))
+    return true;
+  else if (error_if_aux && _problem->getAuxiliarySystem().hasVariable(var_name))
+    mooseError("Variable '",
+               var_name,
+               "' is supposed to be nonlinear for physics '",
+               name(),
+               "' but it is already defined as auxiliary");
+  else
+    return false;
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/physics/PhysicsBase.C]: changes to variableExists

at this point this is more what we want to know, but later on we might want to add a boolean for linear or not

File: modules/navier_stokes/include/physics/WCNSLinearFVFlowPhysics.h [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "WCNSFVFlowPhysicsBase.h"
+#include "WCNSFVTurbulencePhysics.h"
+
+/**
+ * Creates all the objects needed to solve the Navier-Stokes with the SIMPLE algorithm
```
Comment (grmnptr) [inline modules/navier_stokes/include/physics/WCNSLinearFVFlowPhysics.h:16]:
```suggestion
 * Creates all the objects needed to solve the Navier-Stokes equations with the SIMPLE algorithm
```

File: modules/navier_stokes/src/physics/WCNSLinearFVFlowPhysics.C [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+        const std::string bc_type = "LinearFVAdvectionDiffusionOutflowBC";
+        InputParameters params = getFactory().getValidParams(bc_type);
+        params.set<std::vector<BoundaryName>>("boundary") = {outlet_bdy};
+
+        for (const auto d : make_range(dimension()))
+        {
+          params.set<LinearVariableName>("variable") = _velocity_names[d];
+          getProblem().addLinearFVBC(bc_type, _velocity_names[d] + "_" + outlet_bdy, params);
+        }
+      }
```
Comment (grmnptr) [inline modules/navier_stokes/src/physics/WCNSLinearFVFlowPhysics.C:368]:
```suggestion
        const std::string bc_type = "LinearFVAdvectionDiffusionOutflowBC";
        InputParameters params = getFactory().getValidParams(bc_type);
        params.set<std::vector<BoundaryName>>("boundary") = {outlet_bdy};

        for (const auto d : make_range(dimension()))
        {
          params.set<LinearVariableName>("variable") = _velocity_names[d];
          getProblem().addLinearFVBC(bc_type, _velocity_names[d] + "_" + outlet_bdy, params);
        }
```

**Review by grmnptr — CHANGES_REQUESTED** [review]
I was not able to add comments to the file below so I'll just list them here:
- FluidHeatTransferPhyis -> link in text and syntax links might not be correct
- ScalarTransportPhys -> syntax links might not be correct
- SolidHeatTransferPhys -> syntax links might not be correct
- Trubulence -> syntax links might not be correct
- TwoPhaseMixturePhys -> syntax links might not be correct
Looks nice otherwise!

**Comment by GiudGiud** [comment]
good catch I had never looked at these syntax listings

**Comment by GiudGiud** [comment]
Test failures unrelated imo

**Review by grmnptr — APPROVED** [review]
Nice!

**Comment by GiudGiud** [comment]
Thanks for the review!

**Pull request merged**